compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirUselessTypeOperationCallChecker.kt

```diff
@@ -0,0 +1,97 @@
+        if (candidateType is ConeKotlinErrorType) return
+
+        val targetType = expression.conversionTypeRef.coneType.fullyExpandedType(context.session)
+        if (targetType is ConeKotlinErrorType) return
+
+        // x as? Type <=> x as Type?
+        val refinedTargetType =
+            if (expression.operation == FirOperation.SAFE_AS)
+                targetType.withNullability(ConeNullability.NULLABLE, context.session.typeContext)
+            else targetType
+        val shouldCheckForExactType = when (expression.operation) {
+            FirOperation.IS, FirOperation.NOT_IS -> false
+            // FirOperation.AS, FirOperation.SAFE_AS
+            else -> shouldCheckForExactType(expression, context)
+        }
+        if (isRefinementUseless(context, candidateType, refinedTargetType, shouldCheckForExactType)) {
+            when (expression.operation) {
+                FirOperation.IS, FirOperation.NOT_IS -> {
+                    val compileTimeResult = expression.operation == FirOperation.IS
+                    reporter.reportOn(expression.source, FirErrors.USELESS_IS_CHECK, compileTimeResult, context)
+                }
+                // FirOperation.AS, FirOperation.SAFE_AS
+                else -> {
+                    reporter.reportOn(expression.source, FirErrors.USELESS_CAST, context)
+                }
+            }
+        }
+    }
+
+    @Suppress("UNUSED_PARAMETER")
+    private fun shouldCheckForExactType(expression: FirExpression, context: CheckerContext): Boolean {
+        // TODO: differentiate if this expression defines the enclosing thing's type
+        //   e.g.,
+        //   val c1 get() = 1 as Number
+        //   val c2: Number get() = 1 <!USELESS_CAST!>as Number<!>
+        return true
+    }
+
+    private fun isRefinementUseless(
+        context: CheckerContext,
+        candidateType: ConeKotlinType,
+        targetType: ConeKotlinType,
+        shouldCheckForExactType: Boolean,
+    ): Boolean {
+        return if (shouldCheckForExactType)
+            isExactTypeCast(candidateType, targetType)
+        else
+            isUpcast(context, candidateType, targetType)
+    }
+
+    private fun isExactTypeCast(candidateType: ConeKotlinType, targetType: ConeKotlinType): Boolean {
+        return candidateType == targetType &&
+                // See comments at [isUpcast] why we need to check the existence of @ExtensionFunctionType
+                candidateType.isExtensionFunctionType == targetType.isExtensionFunctionType
+    }
+
+    private fun isUpcast(context: CheckerContext, candidateType: ConeKotlinType, targetType: ConeKotlinType): Boolean {
+        if (!AbstractTypeChecker.isSubtypeOf(context.session.typeContext, candidateType, targetType, stubTypesEqualToAnything = false))
+            return false
+
+        if (candidateType.isFunctionalType(context.session) && targetType.isFunctionalType(context.session)) {
```

Comment:
Thanks for the detailed comment on extension function type! This `if` seems redundant though, if we're already checking `isExtensionFunctionType` for both of them.

Reply:
Ah, you're right: this is redundant, as we already checked candidate type is a subtype of a target type, and if one is a functional type, the other should be too. Looks simpler now without `if`!